src/main/java/com/celements/web/service/UrlService.java

```diff
@@ -0,0 +1,103 @@
+package com.celements.web.service;
+
+import java.net.URL;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xwiki.component.annotation.Requirement;
+import org.xwiki.model.reference.AttachmentReference;
+import org.xwiki.model.reference.DocumentReference;
+
+import com.celements.model.access.IModelAccessFacade;
+import com.celements.model.access.exception.DocumentNotExistsException;
+import com.celements.model.context.ModelContext;
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.web.XWikiURLFactory;
+
+public class UrlService implements IUrlService {
+
+  private static Logger LOGGER = LoggerFactory.getLogger(UrlService.class);
+
+  @Requirement
+  private ModelContext context;
+
+  @Requirement
+  private IModelAccessFacade modelAccess;
+
+  @Requirement
+  private ICelementsWebServiceRole celementsweb;
+
+  @Override
+  public String getURL(DocumentReference docRef) {
+    return getURL(docRef, "view");
+  }
+
+  @Override
+  public String getURL(DocumentReference docRef, String action) {
+    return getURL(docRef, action, null);
+  }
+
+  @Override
+  public String getURL(DocumentReference docRef, String action, String queryString) {
+    String spaceName = docRef.getLastSpaceReference().getName();
```

Comment:
add null checks on `docRef`/`attRef` in the new internal `createURLObject`. empty string should be ultimately returned in such cases.